pos-print/src/main/java/com/clouway/pos/print/adapter/db/BackgroundReceiptPrinterService.kt

```diff
@@ -0,0 +1,99 @@
+package com.clouway.pos.print.adapter.db
+
+import com.clouway.pos.print.core.*
+import com.clouway.pos.print.printer.Status
+import com.google.inject.Inject
+import com.mongodb.CursorType
+import com.mongodb.client.MongoCollection
+import org.bson.Document
+import org.slf4j.LoggerFactory
+import kotlin.concurrent.thread
+
+/**
+ * @author Tsvetozar Bonev (tsvetozar.bonev@clouway.com)
+ */
+class BackgroundReceiptPrinterService @Inject constructor(private var repository: ReceiptRepository,
+                                                          private var factory: PrinterFactory,
+                                                          private var notifier: PrintNotifier)
+  : BackgroundReceiptPrinter{
+
+  private val logger = LoggerFactory.getLogger(BackgroundReceiptPrinter::class.java)
+
+  override fun run() {
+    logger.info("Starting background printing service")
+    repository.getCappedCollection().drop()
```

Comment:
This drop will cause issues If app is re-started and some receipts are not printed.

Reply:
All of the receipts are stored in another collection before being passed to the capped one, so none would be lost, the drop is there because a restart would print all of the contained receipts and removing a single receipt from the capped collection isn't allowed.

Reply:
The use of a capped collection here seemed to mimic a normal Java queue as it would need to be dropped at the start and end of the program. On top of that, the mongo tailable cursor was not behaving properly so I switched the implementation for a plain in memory queue that block when asked for the next element and the functionality has remained the same with the bonus of loosing Mongo's capped collection restrictions.

Reply:
Using of in-memory queue is not a good idea as it doesn't allow multiple instances of the pos-print service to be used which removes the redundancy.

Take a look into this one:
http://learnmongodbthehardway.com/schema/queues/

> A queue in the setting of MongoDB, means a collection where inserted documents are kept in an order, either by the creation data of the document or by a priority assigned to the document. Multiple publishers can be inserting new documents into the queue at the same time, but each document can only be delivered to a single listener.


e.g this is what is expected from search service.

Reply:
Multiple instances would all have their own in-memory blocking queue, why would there be a need for a shared queue between the instances?

Reply:
Because both could read the same message and try to print it.

Reply:
>Using of in-memory queue is not a good idea as it doesn't allow multiple instances of the pos-print service to be used which removes the redundancy.

Currently we cannot use many inctances of the pos-print service because we will have and more that one workers, so many pos-print services can write in the same queue receipts for a a same pos printer, and many workers will get them and sen them concurrently to the same printer.

If we would like to use many instances, we should first separate the app to two different apss one for reading and adding receipts to the queue and one that reads the date in the queue.

Other solution could be creating separate qeueu for each pos-printer. 

Using the CappedCollections required the both operations : put in the collection and put in the capped collection to be executed in a transaction(I'm not a hundred 100% sure)

Reply:
Yes, the printing is blocking it for now. Ok, keep it with the in-memory implementation.

Reply:
https://github.com/clouway/pos-print/issues/19

Reply:
Queue separation per IP my do it actually.

Reply:
Also with this in memory queue we can implement logic on the app start, that checks for a not printed receipts and then schedule them in the queue, like that if we restart the app it will load the not printed receipts and trey to print them

Reply:
>Using the CappedCollections required the both operations : put in the collection and put in the capped collection to be executed in a transaction(I'm not a hundred 100% sure)

Yes, it's not atomic now. I suppose it would be better single not capped collection to be used instead as till findAndModify may update it in atomic manner.

> Other solution could be creating separate queue for each pos-printer.

Instances have to take free printer and use it ... this adds a lot of complexity and this is why it would be better to be in-memory. 


As final: the in-memory solution sounds reasonable for know. If we encounter issues then we will may consider implementing another options.

Reply:
I agree with you. At this time the app works syncronous and there a just a few problems.
I think that we should provide and a rest services, to serve the last recepts and a rest service to schedule already saved receipt for printing again(failed receipts). Like that we could provide simple table with that contains last 50 receipts and show button for rescheduling of the failed.